src/openms_gui/source/VISUAL/DIALOGS/ToolsDialog.cpp

```diff
@@ -86,36 +95,40 @@ namespace OpenMS
 
     auto label = new QLabel("TOPP tool:");
     main_grid->addWidget(label, 1, 0);
-    QStringList list;
 
-    // Determine all available tools compatible with the layer_type
+    // Determine all available tools compatible with the layer_type_
     tool_map_ = {
             {FileTypes::Type::MZML, LayerData::DataType::DT_PEAK},
             {FileTypes::Type::MZXML, LayerData::DataType::DT_PEAK},
             {FileTypes::Type::FEATUREXML, LayerData::DataType::DT_FEATURE},
             {FileTypes::Type::CONSENSUSXML, LayerData::DataType::DT_CONSENSUS},
             {FileTypes::Type::IDXML, LayerData::DataType::DT_IDENT}
     };
-    // Get a map of all tools
-    const auto& tools = OpenMS::ToolHandler::getTOPPToolList();
-    for (const auto& tool : tools)
+    // Get a map of all tools and utils
+    const auto& tools = ToolHandler::getTOPPToolList();
+    const auto& utils = ToolHandler::getUtilList();
+    // Extract names of all tools and utils
+    QList<String> keys;
+    for (const auto& pair : tools)
     {
-      const String& tool_name = tool.first;
-      Param p = getParamFromIni_(tool_name);
-      std::vector<LayerData::DataType> tool_types = getTypesFromParam_(p);
-      // Check if tool is compatible with the layer type
-      if (std::find(tool_types.begin(), tool_types.end(), layer_type) != tool_types.end())
-      {
-        list << tool_name.toQString();
-      }
+      keys.push_back(pair.first);
+    }
+    for (const auto& pair : utils)
+    {
+      keys.push_back(pair.first);
     }
+    // Check whether tools/utils are compatible with current layer in background and add all compatible tools/utils
+    // to the tools_combo_ when ready (addEntries_)
+    connect(&param_watcher_, &QFutureWatcher<Param>::finished, this, &ToolsDialog::addEntries_);
+    // Warnings shall be send to TOPPView's LogWindow i.e. the given LogWindow (log_)
+    connect(this, &ToolsDialog::warning, this, &ToolsDialog::logWarning_);
```

Comment:
using signal/slot for internal communication is a bit odd...

Either directly call `logWarning_` instead of emitting, or emit the signal and connect to the logwindow outside of the class.

Reply:
If I'd call `logWarning_` directly each thread would call that function meaning multiple threads access the LogWindow and changing widgets is only permitted for the main thread, no?
Using signals each thread can notify the main thread to log the error so that only the main thread accesses the log widget. That's the reason why I chose to do it that way.

Reply:
ah.. I did not catch that. all good. sorry for the confusion.